src/revlanguage/functions/phylogenetics/ratematrix/Func_InvModel.cpp

```diff
@@ -0,0 +1,153 @@
+    int n = sub_model.getNumberOfStates();
+    vector<double> pi(n,0);
+    for(int m=0; m<f.size(); m++)
+    {
+        auto sub_pi = sub_model.getRootFrequencies(m);
+        for(int l=0;l<pi.size();l++)
+            pi[l] += sub_pi[l] * f[m];
+    }
+    vector<double> er(n*(n-1)/2,0);
+    RevBayesCore::ConcreteTimeReversibleRateMatrix INV(er,pi,{});
+
+    sub_models.push_back( unique_ptr<RevBayesCore::MixtureModel>(new RevBayesCore::UnitMixtureModel(INV)) );
+    
+    auto inv_model = new RevBayesCore::ConcreteMixtureModel(sub_models, fractions);
+    inv_model->setRate(1);
+    return inv_model;
+}
+
+using namespace RevLanguage;
+
+
+/** default constructor */
+Func_InvModel::Func_InvModel( void ) : TypedFunction<MixtureModel>( )
+{
+}
+
+
+/**
+ * The clone function is a convenience function to create proper copies of inherited objected.
+ * E.g. a.clone() will create a clone of the correct type even if 'a' is of derived type 'b'.
+ *
+ * \return A new copy of the process.
+ */
+Func_InvModel* Func_InvModel::clone( void ) const
+{
+    return new Func_InvModel( *this );
+}
+
+
+RevBayesCore::TypedFunction< RevBayesCore::MixtureModel >* Func_InvModel::createFunction( void ) const
+{
+    RevBayesCore::TypedDagNode< RevBayesCore::MixtureModel >* submodel = static_cast<const MixtureModel &>( this->args[0].getVariable()->getRevObject() ).getDagNode();
+    RevBayesCore::TypedDagNode< double >* pInv = static_cast<const RealPos &>( this->args[1].getVariable()->getRevObject() ).getDagNode();
+
+    return RevBayesCore::generic_function_ptr< RevBayesCore::MixtureModel >( InvModelFunc, submodel, pInv );
+}
+
+
+/* Get argument rules */
+const ArgumentRules& Func_InvModel::getArgumentRules( void ) const
+{
+    static ArgumentRules argumentRules = ArgumentRules();
+    static bool          rules_set = false;
+
+    if ( !rules_set )
+    {
+        argumentRules.push_back( new ArgumentRule( "submodel"       , MixtureModel::getClassTypeSpec(), "Sub-rate matrix.", ArgumentRule::BY_CONSTANT_REFERENCE, ArgumentRule::ANY ) );
+        argumentRules.push_back( new ArgumentRule( "pInv"           , RealPos::getClassTypeSpec(), "The fraction of invariable sites.", ArgumentRule::BY_CONSTANT_REFERENCE, ArgumentRule::ANY ) );
+
+        rules_set = true;
+    }
+
+    return argumentRules;
+}
+
+
+const std::string& Func_InvModel::getClassType(void)
+{
+
+    static std::string rev_type = "Func_InvModel";
+
+    return rev_type;
+}
+
+
+/* Get class type spec describing type of object */
+const TypeSpec& Func_InvModel::getClassTypeSpec(void)
+{
+
+    static TypeSpec rev_type_spec = TypeSpec( getClassType(), new TypeSpec( Function::getClassTypeSpec() ) );
+
+    return rev_type_spec;
+}
+
+
+/**
+ * Get the primary Rev name for this function.
+ */
+std::string Func_InvModel::getFunctionName( void ) const
+{
+    // create a name variable that is the same for all instance of this class
+    std::string f_name = "fnInv";
```

Comment:
Same as for `fnGamma`, I would suggest a more descriptive name. Perhaps `fnInvSites`?